Add CF Solve the Maze Problem to Floodfill (#1450)
cpinitiative/usaco-guide: solutions/CF Solve The Maze.mdx

---
id: cf-1365D
source: CF
title: Solve the Maze
author: Daniel Suh
---

[Official Analysis](https://codeforces.com/blog/entry/78504)

## Explanation

The main observation is that if there is a bad person next to a good person, then it is impossible. For example, consider the following:

```
#####
##BG#
###..
####.
```

Because the good person is adjacent to the bad person, the bad person can simply just move right, and now has full access to wherever the good person can move to. It won't matter if the good person can make it, as the bad person can make it as well. Furthermore, replacing a good person with a wall is not allowed, so preventing the movement of the bad person is not allowed. 

> With this observation, the solution is relatively simple. Check for this adjacency, and if it is, print "No." Then, surround the bad people with walls. Floodfill from the end point ($$N - 1,\space M - 1$$), and make sure that all the good people have been visited. You do not need to check if the bad people did not make it, as surrounding the bad people with walls is sufficient. 

Reviewer: ```suggestion
With this observation, the solution is relatively simple. Check for this adjacency, and if it exists, print "No." Otherwise, proceed by surrounding the bad people with walls. Floodfill from the end point $(N - 1,\space M - 1)$, and make sure that all the good people have been visited. You do not need to check if the bad people did not make it, as surrounding the bad people with walls is sufficient. 
```

optional wording changes
